libraries/AP_ESC_Telem/AP_HobbyWing_DataLink.cpp

```diff
@@ -0,0 +1,196 @@
+/*
+  ESC Telemetry for Hobbywing DataLink
+ */
+#include "AP_HobbyWing_DataLink.h"
+
+#if AP_HOBBYWING_DATALINK_ENABLED
+
+#include <AP_HAL/utility/sparse-endian.h>
+#include <AP_HAL/AP_HAL.h>
+
+#include <AP_Math/AP_Math.h>
+
+
+#if 0
+#include <GCS_MAVLink/GCS.h>
+#define Debug(fmt, args ...)  do { gcs().send_text(MAV_SEVERITY_INFO, fmt, ## args); } while (0)
+#else
+#define Debug(fmt, args ...)
+#endif
+
+const AP_Param::GroupInfo AP_HobbyWing_DataLink::var_info[] = {
+    // @Param: _OFS
+    // @DisplayName: HobbyWing DataLink ESC offset
+    // @Description: Used to match incoming ESC telemetry with servo channel number.  If this is 8 then the ESC telemetry coming in from the device plugged into the DataLink device's "1" pins will be matched against ArduPilot's SERVO9 outputs.
+    // @Range: 0 18
+    // @User: Advanced
+    AP_GROUPINFO("_OFS",  10, AP_HobbyWing_DataLink, offset, 0),
+
+    // @Param: _PLS
+    // @DisplayName: HobbyWing DataLink motor pole count
+    // @Description: This allows calculation of true RPM from ESC's eRPM. The default is 14.
+    // @Range: 1 127
+    // @User: Advanced
+    // @RebootRequired: True
+   AP_GROUPINFO("_PLS",  11, AP_HobbyWing_DataLink, motor_poles, 14),
+
+    AP_GROUPEND
+};
+
+void AP_HobbyWing_DataLink::init()
+{
+    uart.begin(115200);
+    uart.set_options(AP_HAL::UARTDriver::OPTION_PULLDOWN_RX);
+
+    // initialise flag values in decoded data:
+    for (uint8_t i=0; i<ARRAY_SIZE(decoded); i++) {
+        decoded[i].rpm = nanf("");
+        decoded[i].voltage = nanf("");
+        decoded[i].phase_current = nanf("");
+        decoded[i].current = nanf("");
+        decoded[i].temperature = nanf("");
+    }
+}
+
+/*
+  update ESC telemetry - note that this is basically a copy of AP_HobbyWing_ESC::read_uart_v345.  Do we want a uart->read_packet(uint8_t *, uint8_t, frame_gap_us)?
+ */
+bool AP_HobbyWing_DataLink::read_uart(uint8_t *_packet, uint8_t packet_len, uint16_t frame_gap_us)
+{
+    bool ret = false;
+
+    const uint32_t n = uart.available();
+
+    if (n < packet_len) {
+        // frame presumably still on its way in
+        return false;
+    }
+
+    const uint64_t this_frame_us = uart.receive_time_constraint_us(packet_len);
+    const uint64_t dt_us = this_frame_us - last_frame_us;
+    if (dt_us < frame_gap_us) {
```

Comment:
where does the frame_gap comes on datalink ?
Can't we just do a simple parsing loop : get header, check for pkt lenght, get crc, compare crc ?

Reply:
header gap is used to strengthen our sanity checks.  Because... "checksum? what checksum?" in some cases, and a pitifully weak 8-bit sum-of-bytes in other cases.

This concept isn't new - I flogged it from tridge's hwing.cpp code.